Add Cone3D

santiagogaray opened the pull request:
Happy New Year @chriswmackey, Sorry it took that long to get back to this, very excited to be back!

As you suggested I did clean up the fork and started a new PR with an initial Cone3D class for your review. 
When you get a chance please let me know your thoughts about constructor parameters as well as the properties included. Also feel free to add any functionality that may be useful to have in this geometry object.

chriswmackey commented:
Thanks @santiagogaray . I will review this as soon as I get the chance (hopefully today).

chriswmackey commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+        axis: A Vector3D representing the direction of the cone.
+        angle: A degree in radians representing the half angle between
+            the axis and the surface.
+        height: A number representing the height of the cone.
```
I was under the initial impression that the magnitude of the axis vector gives you the height of the cone but I see that you have a completely separate property for height here. Am I correct in assuming that the magnitude of the axis vector is meaningless in the current setup? If so, I think we should change it and we should get rid of this height property in favor of using the vector's magnitude. The lighter and less redundant we can keep these objects's properties, the better.

santiagogaray replied:
Yes, I was about to make axis normalized to be less redundant, mostly because I was not sure if it was okay to combine two cone math elements in one python element. But I agree we need to decide in favor of what keep things lighter, and I see now how it does simplify a lot of things throughout the code.

chriswmackey commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+        return Cone3D(self.origin.scale(factor, origin),
+                      self.axis,
+                      self.angle,
+                      self.height)
```
You will have to scale the height of the cone in addition to moving the vertex. Or rather, you'll have to scale the axis vector (after you make the change noted above).

santiagogaray replied:
Done!

chriswmackey commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+        return 'Cone3D (vertex {}) (axis {}) (angle {}) (height {}))'.format(self.vertex,
+                                                                             self.axis,
+                                                                             self.angle,
+                                                                             self.height)
```
I'd recommend putting all of the properties inside the `format` bracket on one line. As long as you indent it from the one above it, this should work fine and be a bit more readable.

santiagogaray replied:
Yep, my bad!

chriswmackey commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+                'axis': self.axis.to_array(), 'angle': self.angle, 'height': self.height}
+
+    def __copy__(self):
+        return Cone3D(self.vertex, self.axis, self.angle, self.height)
```
In addition to overwriting the hidden `__copy__` method, there are now a few other hidden methods that we overwrite on all geometry objects, which make it easier for us to tell if a given geometry object is equal to another.  Please add these methods to this class and you can see an example for all of them here in the Arc3D class:
https://github.com/ladybug-tools/ladybug-geometry/blob/master/ladybug_geometry/geometry3d/arc.py#L329-L340

santiagogaray replied:
Yes, I saw those throughout the geometry package. Nice addition!

santiagogaray commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+
+    Args:
+        vertex: A Point3D representing the vertex of the cone.
+        axis: A Vector3D representing the direction of the cone.
```
Good point, hope this is clear now.

santiagogaray commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+    """Cone3D object.
+
+    Args:
+        vertex: A Point3D representing the vertex of the cone.
```
Done!

santiagogaray commented on ladybug_geometry/geometry3d/cone.py:
```diff
@@ -0,0 +1,191 @@
+        vertex: A Point3D representing the vertex of the cone.
+        axis: A Vector3D representing the direction of the cone.
+        angle: A degree in radians representing the half angle between
+            the axis and the surface.
```
My bad, corrected!

santiagogaray commented:
Thanks @chriswmackey for taking the time to review everything. Feel free to take a look at the changes addressing your comments. I also added some cone tests for your review.
